src/vars.c

```diff
@@ -367,19 +367,19 @@ static Obj EvalIsbHVar(Expr expr)
 **
 **  'PrintAssHVar' prints the higher variable assignment statement <stat>.
 */
-static void PrintAssHVar(Stat stat)
+static void PrintAssHVar(Obj body, Stat stat)
 {
     Pr( "%2>", 0L, 0L );
-    Pr("%H", (Int)NAME_HVAR(READ_STAT(stat, 0)), 0L);
+    Pr("%H", (Int)NAME_HVAR(READ_STAT_IN_BODY(body, stat, 0)), 0L);
```

Comment:
Oops, I just realized this call to `NAME_HVAR` here and in a few other places is a serious problem: it requires the environment of the function object; and that is *not* available from the body alone.... *sigh*.

So it seems we either need to pass in the `T_FUNCTION` bag, not just the `T_BODY` bag; (which is a bit sad, because we only need the body in all other print functions, as far as I can tell). Or, we need to redesign how hvar references work: instead of walking about the lvars chain, we could also store any relevant `T_LVARS` resp. `T_HVARS` inside the `values` plist of the body; then adjust the code writing to HVARS to go through that. I am not quite sure whether that's a terrific or terrible idea, though...

Reply:
The lexical environment of a function body (ie the enclosing function bodies) should logically be an attribute of the body itself, not the function and is enough to find names, though of course not values, of higher variables (debug variables are whole  other problem). 

If we adjusted our data structures to reflect this, then a function would lead to two chains of pointers -- one from functions to the local variables bags of the function calls in which they were created, and one from bodies to the lexically enclosing bodies. This is redundant, with all the usual risks of inconsistency, but without it, a function body with higher variable references is meaningless when taken out of context.

Another option is just to cache the names of the higher variables in the function body as a convenience.